[TOC]

# EasyNanoMeta:A pipeline for nanopore long-read metagenomic analysis

With the widespread application of nanopore sequencing in microbiology, there has been a significant increase in the generation of long-read sequencing data. This poses a challenge in effectively mining valuable biological information from the vast amount of nanopore sequencing data using bioinformatics tools. In this review, we present the advantages and pitfalls of long-read nanopore sequencing technologies in microbial genome and microbiome research, systematically summarize the bioinformatics analysis methods for nanopore long-read metagenomics and validate the practical implementation of the analysis workflow using various simulated and experimental metagenomic datasets. In addition, we conducted a comprehensive evaluation of the overall performance and computational requirement of various bioinformatic tools.

## Workflow overview

![image](https://github.com/P-kai/EasyNanoMeta/blob/main/Figures/figure2.png)

The main analysis pipeline for nanopore microbial sequences. This comprehensive statement encompasses the various steps involved in the analysis process, along with the corresponding tools employed. The initial step in analyzing nanopore sequencing data involves data quality control, including basecalling, filtering and host sequences removal. Presently, several tools can be utilized for the analysis of full-length 16S sequencing data obtained through nanopore sequencing. Emu is the most reliable and user-friendly option in this regard. For bacterial genome analysis, accurate and complete bacterial genomes can be generated via two distinct strategies. The first approach entails performing long-read assembly and subsequently refining the assembly using high-accuracy short-read data. The second approach involves hybrid assembly, which combines both long-read and short-read data. Nanopore metagenomic data analysis is more complex than 16S data and bacterial genome data. Assembly-free long-read metagenomic profiling can quickly estimate the abundance of species and functional genes employing few computing resources. In comparison, assembly-based nanopore metagenomic data analysis enable us to study more high-quality MAGs and their functions.
